// cc-cedict/dist/build.js
import https from 'https';
import fs from 'fs';
import unzipper from 'unzipper';
import { fileURLToPath } from 'url';
import path from 'path';
const cedictUrl = 'https://www.mdbg.net/chinese/export/cedict/cedict_1_0_ts_utf-8_mdbg.zip';
const dataPath = path.join(path.dirname(fileURLToPath(import.meta.url)), '../data');
const PATHS = {
    status: path.join(dataPath, 'status.json'),
    zipFile: path.join(dataPath, 'cedict.zip'),
    all: path.join(dataPath, 'all.js'),
    traditional: path.join(dataPath, 'traditional.js'),
    simplified: path.join(dataPath, 'simplified.js'),
};
export const REGEX = {
    line: /\/(.*)/s,
    variant_of: /(variant of (([\p{Unified_Ideograph}\u3006\u3007][\ufe00-\ufe0f\u{e0100}-\u{e01ef}]?){1,})?(\|([\p{Unified_Ideograph}\u3006\u3007][\ufe00-\ufe0f\u{e0100}-\u{e01ef}]?){1,})?(\[([^\]]*))?)/gmu,
    classifiers: /(CL:((([\p{Unified_Ideograph}\u3006\u3007][\ufe00-\ufe0f\u{e0100}-\u{e01ef}]?){1,})?(\|([\p{Unified_Ideograph}\u3006\u3007][\ufe00-\ufe0f\u{e0100}-\u{e01ef}]?){1,})?(\[([^\]]*)\]),?)+)/gmu,
    pinyin: /([A-Za-z\:]+[0-9])/g,
};
// Check if we should show progress messages based on npm log level
const shouldShowProgress = () => {
    const logLevel = process.env.npm_config_loglevel;
    return logLevel !== 'silent' && logLevel !== 'error';
};
// Log informational messages (respects npm log level)
const info = (msg) => {
    if (shouldShowProgress()) {
        console.error(`cc-cedict: ${msg}`);
    }
};
// Log error messages (always shown, with package prefix)
const error = (msg, ...args) => {
    console.error(`cc-cedict: ${msg}`, ...args);
};
// Utility to ensure a directory exists
const ensureDirectoryExists = (dir) => {
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
    }
};
// Downloads a file from the given URL
const downloadFile = (url, dest) => {
    return new Promise((resolve, reject) => {
        const file = fs.createWriteStream(dest);
        https
            .get(url, (response) => {
            if (response.statusCode !== 200) {
                reject(new Error(`Failed to download '${url}' (status code: ${response.statusCode})`));
                return;
            }
            response.pipe(file);
        })
            .on('error', (err) => {
            reject(err);
        });
        file.on('finish', () => {
            file.close(() => resolve());
        });
        file.on('error', (err) => {
            fs.unlink(dest, () => reject(err)); // Cleanup if there's an error
        });
    });
};
// Push a parsed variant into the data structure
const pushVariantTo = (val, orig, lineToNum, arr, field // 0 for traditional, 1 for simplified
) => {
    if (!val)
        return;
    if (!orig)
        return;
    const valKey = `${val[0]}_${val[1]}_${val[2]}`;
    const hanzi = orig[field];
    const pinyin = orig[2];
    const valHanzi = val[field];
    const valPinyin = val[2];
    if (!hanzi || // no hanzi in original value
        (orig !== val && !arr[hanzi] && (!pinyin || !valHanzi || !valPinyin)) || // variant for non-existing field in array lacking necessary data
        !valHanzi || // no hanzi in new value
        !valPinyin // no pinyin in new value
    )
        return;
    if (!pinyin)
        return;
    arr[hanzi] ??= {};
    arr[hanzi][pinyin] ??= [[], []];
    const indexEntry = arr[hanzi][pinyin];
    indexEntry[val !== orig ? 1 : 0].push(lineToNum[valKey]);
};
// Extract and process the ZIP file
const extractAndProcessZip = async (zipPath) => {
    try {
        const directory = await unzipper.Open.file(zipPath);
        const cedictFile = directory.files.find((f) => f.path === 'cedict_ts.u8');
        if (!cedictFile) {
            error('✗ cedict_ts.u8 not found in the ZIP file');
            return;
        }
        const cedictData = (await cedictFile.buffer()).toString('utf8');
        const lines = cedictData.split('\n');
        const variantQueue = [];
        const lineToNumSet = new Set(); // Optimized: Use Set for duplicate detection
        const lineToNum = {};
        const all = [];
        const traditional = {};
        const simplified = {};
        // Optimized: Create lookup tables for variants and classifiers
        const variantLookupMap = new Map();
        const classifierLookupMap = new Map();
        const variantLookup = [];
        const classifierLookup = [];
        // Helper to add to lookup table
        const addToLookup = (tuple, lookupMap, lookupArray) => {
            const key = `${tuple[0]}_${tuple[1]}_${tuple[2]}`;
            let idx = lookupMap.get(key);
            if (idx === undefined) {
                idx = lookupArray.length;
                lookupArray.push([tuple[0], tuple[1], tuple[2] || '']);
                lookupMap.set(key, idx);
            }
            return idx;
        };
        for (const line of lines) {
            const parsedLine = parseLine(line);
            if (!parsedLine || !parsedLine[0] || !parsedLine[1] || !parsedLine[2])
                continue;
            const key = `${parsedLine[0]}_${parsedLine[1]}_${parsedLine[2]}`;
            if (parsedLine[4].length) {
                variantQueue.push(parsedLine);
            }
            // Optimized: Use Set for faster duplicate detection
            if (lineToNumSet.has(key))
                throw key;
            lineToNumSet.add(key);
            lineToNum[key] = all.length;
            // Convert variant tuples to indices
            const variantIndices = parsedLine[4].map(tuple => addToLookup(tuple, variantLookupMap, variantLookup));
            // Convert classifier tuples to indices
            const classifierIndices = parsedLine[5].map(tuple => addToLookup(tuple, classifierLookupMap, classifierLookup));
            // Optimized: Store single meaning as string, multiple as array
            const meanings = parsedLine[3].length === 1 ? parsedLine[3][0] : parsedLine[3];
            // Create optimized entry
            const optimizedEntry = [
                parsedLine[0],
                parsedLine[1],
                parsedLine[2],
                meanings,
                variantIndices,
                classifierIndices
            ];
            pushVariantTo(parsedLine, parsedLine, lineToNum, traditional, 0);
            pushVariantTo(parsedLine, parsedLine, lineToNum, simplified, 1);
            all.push(optimizedEntry);
        }
        for (const parsedLine of variantQueue) {
            if (!parsedLine)
                continue;
            for (const original of parsedLine[4]) {
                pushVariantTo(parsedLine, original, lineToNum, traditional, 0);
                pushVariantTo(parsedLine, original, lineToNum, simplified, 1);
            }
        }
        // Create the optimized data structure for all.js (entries + lookup tables only)
        const allData = {
            all,
            variantLookup,
            classifierLookup
        };
        // Write optimized data files
        fs.writeFileSync(PATHS.all, `export default ${JSON.stringify(allData)}`);
        fs.writeFileSync(PATHS.traditional, `export default ${JSON.stringify(traditional)}`);
        fs.writeFileSync(PATHS.simplified, `export default ${JSON.stringify(simplified)}`);
    }
    catch (err) {
        error('✗ Error processing ZIP file:', err instanceof Error ? err.stack || err : err);
    }
};
// Parse a single line of the dictionary
export const parseLine = (line) => {
    line = line.trim();
    if (!line || line.startsWith('#'))
        return null;
    const splitLine = line.split(REGEX.line);
    if (!splitLine || splitLine.length < 2)
        return null;
    try {
        const parsed = parseMeanings(splitLine[1]);
        const [chars, pinyinPart] = splitLine[0].split('[');
        if (!chars || !pinyinPart)
            return null;
        const [traditional, simplified] = chars.trim().split(' ');
        const pinyin = pinyinPart.split(']')[0];
        if (!traditional || !simplified || !pinyin)
            return null;
        return [traditional, simplified, pinyin, parsed.meanings, parsed.variant_of, parsed.classifiers];
    }
    catch (e) {
        error('✗ Error parsing line:', e instanceof Error ? e.stack || e : e);
        return null;
    }
};
// Parse meanings and variants
export const parseMeanings = (input) => {
    const result = { meanings: [], variant_of: [], classifiers: [] };
    const variantMap = {};
    const classifierMap = {};
    for (const meaning of input.replace('\r', '').split('/')) {
        const trimmed = meaning.trim();
        if (!trimmed)
            continue;
        const variantMatches = trimmed.match(REGEX.variant_of);
        let skipMeaning = false;
        if (variantMatches) {
            if (variantMatches[0] === trimmed)
                skipMeaning = true;
            const variant = variantMatches[0].substring(11);
            const parsed = parseVariant(variant);
            if (!parsed)
                continue;
            const key = Object.values(parsed).join('');
            if (!variantMap[key]) {
                variantMap[key] = true;
                result.variant_of.push(parsed);
            }
        }
        const classifierMatches = trimmed.match(REGEX.classifiers);
        if (classifierMatches) {
            if (classifierMatches[0] === trimmed)
                skipMeaning = true;
            const classifiers = classifierMatches[0].substring(3).split(',');
            for (const classifier of classifiers) {
                const parsed = parseVariant(classifier);
                if (!parsed)
                    continue;
                const key = Object.values(parsed).join('');
                if (!classifierMap[key]) {
                    classifierMap[key] = true;
                    result.classifiers.push(parsed);
                }
            }
        }
        if (!skipMeaning) {
            result.meanings.push(trimmed);
        }
    }
    return result;
};
// Parse a variant definition
export const parseVariant = (input) => {
    if (!input || !input.length)
        return;
    const [chars, pinyinPart] = input.split('[');
    const [simplified, traditional] = chars.split('|');
    const pinyin = pinyinPart?.match(REGEX.pinyin)?.join(' ') || null;
    return [simplified, traditional || simplified, pinyin];
};
// Main entry point
const main = async () => {
    try {
        ensureDirectoryExists(dataPath);
        info('↓ Downloading CC-CEDICT dictionary...');
        let success = false;
        for (let attempt = 1; attempt <= 3; attempt++) {
            try {
                await downloadFile(cedictUrl, PATHS.zipFile);
                success = true;
                break;
            }
            catch (err) {
                error(`✗ Download attempt ${attempt} failed:`, err);
            }
        }
        try {
            if (!success)
                throw 'Failed to download CC-CEDICT';
            info('↻ Parsing dictionary data (this may take a moment)...');
            await extractAndProcessZip(PATHS.zipFile);
            info('↻ Cleaning up...');
            fs.unlinkSync(PATHS.zipFile);
            fs.writeFileSync(PATHS.status, JSON.stringify({ updated_at: new Date().toISOString() }));
        }
        catch (e) {
            error(`⚠ ${e}. Using fallback CC-CEDICT data processed at package upload time...`);
            fs.unlinkSync(PATHS.zipFile);
            return;
        }
        info('✅ Setup complete!');
    }
    catch (err) {
        error('✗ Error in main execution:', err instanceof Error ? err.stack || err : err);
    }
};
// Only run main() when this file is executed directly (not when imported)
if (import.meta.url === `file://${process.argv[1]}`) {
    main();
}
